source/vir/src/interpreter.rs

```diff
@@ -0,0 +1,1438 @@
+        64 => BigInt::from_u64(u as u64),
+        128 => BigInt::from_u128(u as u128),
+        _ => panic!("Unexpected fixed-width integer type U({})", width),
+    }
+    .unwrap()
+}
+
+/// Truncate an i128 to a fixed width BigInt
+fn i128_to_fixed_width(i: i128, width: u32) -> BigInt {
+    match width {
+        8 => BigInt::from_i8(i as i8),
+        16 => BigInt::from_i16(i as i16),
+        32 => BigInt::from_i32(i as i32),
+        64 => BigInt::from_i64(i as i64),
+        128 => BigInt::from_i128(i as i128),
+        _ => panic!("Unexpected fixed-width integer type U({})", width),
+    }
+    .unwrap()
+}
+
+/// Displays data for profiling/debugging the interpreter
+fn display_perf_stats(state: &State) {
+    if state.perf {
+        state.log(format!("\n{}\nPerformance Stats\n{}\n", "*".repeat(80), "*".repeat(80)));
+        state.log(format!("Performed {} interpreter iterations", state.iterations));
+        if state.enable_simplified_cache {
+            let sum = state.ptr_hits + state.ptr_misses;
+            let hit_perc = 100.0 * (state.ptr_hits as f64 / sum as f64);
+            state.log(format!(
+                "Simplified cache had {} hits out of {} ({:.1}%)",
+                state.ptr_hits, sum, hit_perc
+            ));
+        } else {
+            state.log("Simplified cache was disabled".to_string());
+        }
+        if state.enable_cache {
+            let sum = state.cache_hits + state.cache_misses;
+            let hit_perc = 100.0 * (state.cache_hits as f64 / sum as f64);
+            state.log(format!(
+                "Call result cache had {} hits out of {} ({:.1}%)",
+                state.cache_hits, sum, hit_perc
+            ));
+
+            let mut cache_stats: Vec<(&Fun, usize)> =
+                state.cache.iter().map(|(fun, vec)| (fun, vec.len())).collect();
+            cache_stats.sort_by(|a, b| b.1.cmp(&a.1));
+            for (fun, calls) in &cache_stats {
+                state.log(format!("{:?} cached {} distinct invocations", fun.path, calls));
+            }
+        } else {
+            state.log("Function-call cache was disabled".to_string());
+        }
+        state.log(format!("\nRaw call numbers:"));
+        let mut fun_call_stats: Vec<(&Fun, _)> = state.fun_calls.iter().collect();
+        fun_call_stats.sort_by(|a, b| b.1.cmp(&a.1));
+        for (fun, count) in fun_call_stats {
+            state.log(format!("{:?} called {} times", fun.path, count));
+        }
+    }
+}
+
+/***********************************************
+ * Special handling for interpreting sequences *
+ ***********************************************/
+
+enum SeqFn {
+    Empty,
+    New,
+    Push,
+    Update,
+    Subrange,
+    Add,
+    Len,
+    Index,
+    ExtEqual,
+    Last,
+}
+
+/// Identify sequence functions for which we provide custom interpretation
+fn is_sequence_fn(fun: &Fun) -> Option<SeqFn> {
+    use SeqFn::*;
+    match path_as_rust_name(&fun.path).as_str() {
```

Comment:
Can you add a TODO to use a more robust mechanism for this matching? We may want to use either diagnostic_item or our own attribute to connect these (so that if they ever move they don't get detached from the semantics that we associate here). We should do this for other internal functions, too.

Reply:
Yep, done.